atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/metrics/LastSweptTimestampUpdaterTest.java

```diff
@@ -0,0 +1,192 @@
+        verify(executorService, times(1))
+                .scheduleWithFixedDelay(any(), eq(REFRESH_MILLIS), eq(REFRESH_MILLIS), eq(TimeUnit.MILLISECONDS));
+    }
+
+    @Test
+    public void scheduledTaskDoesNotInteractWithMetricsOrQueueBeforeDelayIsElapsed() {
+        lastSweptTimestampUpdater.schedule(REFRESH_MILLIS);
+        executorService.tick(REFRESH_MILLIS - 1, TimeUnit.MILLISECONDS);
+        verifyNoMoreInteractions(queue, metrics);
+    }
+
+    @Test
+    public void scheduledTaskInteractsWithMetricsAndQueueAsExpectedAfterOneDelay() {
+        stubWithRealisticReturnValues();
+        lastSweptTimestampUpdater.schedule(REFRESH_MILLIS);
+        executorService.tick(REFRESH_MILLIS, TimeUnit.MILLISECONDS);
+
+        verify(queue, atLeast(1)).getNumShards();
+        verify(queue, atLeast(1)).getLastSweptTimestamps(conservativeShardAndStrategySet);
+        verify(queue, atLeast(1)).getLastSweptTimestamps(thoroughShardAndStrategySet);
+
+        for (int shard = 0; shard < shards; shard++) {
+            verify(metrics, times(1)).updateProgressForShard(ShardAndStrategy.conservative(shard), CONS_TS);
+            verify(metrics, times(1)).updateProgressForShard(ShardAndStrategy.thorough(shard), THOR_TS);
+        }
+        verifyNoMoreInteractions(queue, metrics);
+    }
+
+    @Test
+    public void scheduledTaskInteractsWithMetricsAndQueueAsExpectedAfterMultipleDelays() {
+        stubWithRealisticReturnValues();
+        lastSweptTimestampUpdater.schedule(REFRESH_MILLIS);
+        executorService.tick(TICK_COUNT * REFRESH_MILLIS, TimeUnit.MILLISECONDS);
+
+        verify(queue, atLeast(TICK_COUNT)).getNumShards();
+        verify(queue, atLeast(TICK_COUNT)).getLastSweptTimestamps(conservativeShardAndStrategySet);
+        verify(queue, atLeast(TICK_COUNT)).getLastSweptTimestamps(thoroughShardAndStrategySet);
+
+        for (int shard = 0; shard < shards; shard++) {
+            verify(metrics, times(TICK_COUNT)).updateProgressForShard(ShardAndStrategy.conservative(shard), CONS_TS);
+            verify(metrics, times(TICK_COUNT)).updateProgressForShard(ShardAndStrategy.thorough(shard), THOR_TS);
+        }
+        verifyNoMoreInteractions(queue, metrics);
+    }
+
+    @Test
+    public void scheduledTaskKeepsRunningAfterUpdateProgressForShardFails() {
+        when(queue.getLastSweptTimestamps(anySet())).thenReturn(Collections.singletonMap(CONS_SHARD, CONS_TS));
+        lastSweptTimestampUpdater.schedule(REFRESH_MILLIS);
+
+        doThrow(RuntimeException.class)
+                .doThrow(RuntimeException.class)
+                .doThrow(Error.class)
+                .doThrow(Error.class)
+                .doNothing()
+                .when(metrics)
+                .updateProgressForShard(any(), anyLong());
+
+        executorService.tick(3 * REFRESH_MILLIS, TimeUnit.MILLISECONDS);
+
+        verify(metrics, times(2 * 3)).updateProgressForShard(CONS_SHARD, CONS_TS);
```

Comment:
It took me a little while to get this - it's because even for thorough, you're returning the cons shard and cons_ts.

Might be a little easier to parse by mocking the getLastSweptTimestamps part properly e.g just return the empty set for thorough